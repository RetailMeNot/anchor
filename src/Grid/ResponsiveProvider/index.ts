export {
    ResponsiveProvider,
    ResponsiveContext,
} from './ResponsiveProvider.component';
