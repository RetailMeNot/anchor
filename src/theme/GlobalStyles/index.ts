export { GlobalCSS } from './Globals.styles';
export { NormalizeCSS } from './Normalize.styles';
