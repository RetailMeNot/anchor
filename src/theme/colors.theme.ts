export default {
  neutrals: {
    charcoal: {
      base: '#323232',
      light: '#595959',
      dark: '#000000',
    },
    ash: {
      base: '#A6A6A6',
      light: '#D3D3D3',
      dark: '#808080',
    },
    silver: {
      base: '#F1F1F1',
      light: '#FAFAFA',
      dark: '#E7E7E7',
    },
    lavenderWhispers: {
      base: '#F7F5FC',
      light: '#FBFAFF',
      dark: '#F1EEFA',
    },
    white: {
      base: '#FFFFFF',
    },
  },
  main: {
    grapePurchase: {
      base: '#602D6C',
      light: '#91609C',
      dark: '#4B2354',
    },
    rx: {
      base: '#F23466',
      light: '#FF6B9C',
      dark: '#CC2D52',
    },
    genie: {
      base: '#4C18F3',
      light: '#7855D9',
      dark: '#42199A',
    },
  },
  accent: {
    savvyCyan: {
      base: '#0998D6',
      light: '#66CCFF',
      dark: '#0074A6',
    },
    tealBreaker: {
      base: '#00CCCD',
      light: '#00E0D1',
      dark: '#00BDBD',
    },
  },
  secondary: {
    dealEnvy: {
      base: '#14A382',
      light: '#00CDAD',
      dark: '#268068',
    },
    flashPink: {
      base: '#DF206C',
      light: '#E665A1',
      dark: '#B81D5B',
    },
    cyberMango: {
      base: '#EB663D',
      light: '#FA9F5A',
      dark: '#C74E30',
    },
    goldMine: {
      base: '#F5A623',
      light: '#FFC761',
      dark: '#CC861D',
    },
    fireSale: {
      base: '#EA4545',
      light: '#F16667',
      dark: '#CC3B3B',
    },
  },
  alert: {
    error: {
      base: '#CC3B3B',
    },
    success: {
      base: '#268068',
    },
  },
  rx: {
    purple: {
      base: '#A678B5',
    },
  },
};
