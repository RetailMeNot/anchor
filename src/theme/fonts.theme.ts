// VARIABLES
import variables from './variables.theme';

const { baseFontSize, fontFamily } = variables;

export default {
  baseFontSize,
  fontFamily,
  paragraph: {
    base: {
      fontSize: `${baseFontSize}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    informational: {
      fontSize: `${baseFontSize *  0.875}px`,
      lineHeight: `${baseFontSize * 1.125}px`,
      fontWeight: 'normal',
    },
  },
  navigation: {
    base: {
      fontSize: `${baseFontSize * .75}px`,
      lineHeight: `${baseFontSize}px`,
      fontWeight: '500',
    },
    menu: {
      fontSize: `${baseFontSize * .875}px`,
      lineHeight: `${baseFontSize}px`,
      small: {
        base: {
          fontSize: `${baseFontSize * 0.75}px`,
          fontWeight: '500',
        },
        active: {
          fontWeight: '700',
        },
      },
      large: {
        base: {
          fontWeight: '600',
        },
        active: {
          fontWeight: '700',
        },
      },
    },
  },
  heading: {
    h4: {
      fontSize: `${baseFontSize * 1.5}px`,
      lineHeight: `${baseFontSize * 2}px`,
      fontWeight: 'normal',
    },
    h3: {
      fontSize: `${baseFontSize * 1.75}px`,
      lineHeight: `${baseFontSize * 2}px`,
      fontWeight: 'normal',
    },
    h2: {
      fontSize: `${baseFontSize * 2}px`,
      lineHeight: `${baseFontSize * 2.5}px`,
      fontWeight: 'normal',
    },
    h1: {
      fontSize: `${baseFontSize * 2.25}px`,
      lineHeight: `${baseFontSize * 2.5}px`,
      fontWeight: 'normal',
    },
  },
  typography: {
    caption: {
      fontSize: `${baseFontSize * 0.75}px`,
      lineHeight: `${baseFontSize}px`,
      fontWeight: '500',
    },
    ui: {
      fontSize: `${baseFontSize * 0.875}px`,
      lineHeight: `${baseFontSize * 1.125}px`,
      fontWeight: 'normal',
    },
    body: {
      fontSize: `${baseFontSize}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    subHeading: {
      fontSize: `${baseFontSize * 1.125}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    subHeadingLarge: {
      fontSize: `${baseFontSize * 1.25}px`,
      lineHeight: `${baseFontSize * 1.5}px`,
      fontWeight: 'normal',
    },
    header: {
      small: {
        fontSize: `${baseFontSize * 1.5}px`,
        lineHeight: `${baseFontSize * 2}px`,
        fontWeight: 'normal',
      },
      medium: {
        fontSize: `${baseFontSize * 1.75}px`,
        lineHeight: `${baseFontSize * 2}px`,
        fontWeight: 'normal',
      },
      large: {
        fontSize: `${baseFontSize * 2}px`,
        lineHeight: `${baseFontSize * 2.5}px`,
        fontWeight: 'normal',
      },
      extraLarge: {
        fontSize: `${baseFontSize * 2.25}px`,
        lineHeight: `${baseFontSize * 2.5}px`,
        fontWeight: 'normal',
      },
    },
    display: {
      small: {
        fontSize: `${baseFontSize * 2.5}px`,
        lineHeight: `${baseFontSize * 3}px`,
        fontWeight: 'normal',
      },
      medium: {
        fontSize: `${baseFontSize * 3}px`,
        lineHeight: `${baseFontSize * 3.25}px`,
        fontWeight: 'normal',
      },
      large: {
        fontSize: `${baseFontSize * 3.5}px`,
        lineHeight: `${baseFontSize * 4}px`,
        fontWeight: 'normal',
      },
      extraLarge: {
        fontSize: `${baseFontSize * 4}px`,
        lineHeight: `${baseFontSize * 4.5}px`,
        fontWeight: 'normal',
      },
    },
  },
};
