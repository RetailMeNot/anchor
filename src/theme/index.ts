export { default as colors } from './colors.theme';
export { default as fonts } from './fonts.theme';
export { default as sizes } from './sizes.theme';
export { default as variables } from './variables.theme';

export { NormalizeCSS, GlobalCSS } from './GlobalStyles';
