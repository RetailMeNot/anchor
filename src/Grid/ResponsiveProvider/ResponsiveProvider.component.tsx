// Vendor
import * as React from 'react';
import styled, { withTheme } from '@xstyled/styled-components';
import { debounce } from 'ts-debounce';
// COMPONENTS
import { RootTheme } from '../../theme';
// UTILS
import { BreakpointType, getBreakpointKey, sortBreakpoints } from '../utils';

const StyledDebug = styled('div')`
    padding: 1rem 2rem;
    z-index: 1000000;
    background: black;
    color: white;
    position: fixed;
    bottom: 1rem;
    right: 1rem;
    font-family: mono;
`;
StyledDebug.displayName = 'StyledDebug';

interface ResponsiveContextProps {
    breakpoints: BreakpointType[];
    current: string;
    innerWidth: number;
}

export const ResponsiveContext = React.createContext<ResponsiveContextProps>({
    breakpoints: [],
    current: '',
    innerWidth: 0,
});

interface ResponsiveProviderProps {
    breakpoints?: BreakpointType;
    children?: any;
    debug?: boolean;
    theme?: object;
    windowDep?: boolean; // Solely for testing. Forces the component to not recognize the window.
}

class BasicResponsiveProvider extends React.PureComponent<
    ResponsiveProviderProps
> {
    state: ResponsiveContextProps;

    hasWindow: boolean;

    constructor(props: any) {
        super(props);

        const hasWindow =
            props.windowDep === false ? false : typeof window !== 'undefined';
        let breakpoints;

        // Priority is props > ThemeProvider > RootTheme
        if (props.breakpoints !== undefined) {
            breakpoints = props.breakpoints;
        } else if (props.theme !== undefined) {
            breakpoints = props.theme.breakpoints;
        } else {
            breakpoints = RootTheme.breakpoints;
        }

        const sortedBreakpoints = sortBreakpoints(breakpoints);
        const innerWidth = hasWindow ? window.innerWidth : 0;

        this.state = {
            breakpoints: sortedBreakpoints,
            current: getBreakpointKey(innerWidth, sortedBreakpoints),
            innerWidth,
        };
        this.handleResize = debounce(this.handleResize.bind(this), 100);
        this.hasWindow = hasWindow;
    }

    handleResize() {
        if (this.hasWindow) {
            this.setState({
                current: getBreakpointKey(
                    window.innerWidth,
                    this.state.breakpoints
                ),
                innerWidth: window.innerWidth,
            });
        }
    }

    componentDidMount() {
        if (this.hasWindow) {
            window.addEventListener('resize', this.handleResize);
        }
    }

    componentWillUnmount() {
        if (this.hasWindow) {
            window.removeEventListener('resize', this.handleResize);
        }
    }

    render() {
        const { breakpoints, current, innerWidth } = this.state;
        const { debug } = this.props;

        return (
            <ResponsiveContext.Provider
                value={{ breakpoints, current, innerWidth }}
            >
                {debug && (
                    <StyledDebug>
                        {current} | w:
                        {innerWidth}px
                    </StyledDebug>
                )}

                {this.props.children}
            </ResponsiveContext.Provider>
        );
    }
}

export const ResponsiveProvider = withTheme(BasicResponsiveProvider);
