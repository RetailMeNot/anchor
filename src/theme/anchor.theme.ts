// ANCHOR THEME
import { radii } from './sizes.theme';
import { MODAL_THEME_KEY, MODAL_THEME } from '../Modal';

// Copied from Adapter.StandardBreakpoints mins
// todo: dedup- they should use a single source
const breakpoints = {
    sm: '576px',
    md: '768px',
    lg: '992px',
    xl: '1200px',
};

interface AnchorTheme {
    colors?: any;
    breakpoints?: any;
    radii?: any;
    [key: string]: any;
}

export const anchorTheme: AnchorTheme = {
    breakpoints,
    radii,
    [MODAL_THEME_KEY]: MODAL_THEME,
};
