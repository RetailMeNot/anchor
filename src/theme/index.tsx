// Theme Files
import { typography } from './typography.theme';
import { ColorsTheme } from './colors.theme';
import { ThemeProvider as XstyledThemeProvider } from '@xstyled/styled-components';
import styled from '@xstyled/styled-components';
import { space as spaceStyles, SpaceProps } from '@xstyled/system';
import { BUTTON_KEY, BUTTON_THEME } from '../Button';
import { INPUT_KEY, INPUT_THEME } from '../Form/Input/utils';
import React from 'react';
import {Button} from '../Button';
import classNames from 'classnames';

export const RootTheme = {
    typography,
    radii: { none: '0', base: '4px', modal: '8px', circular: '1000px' },
    [BUTTON_KEY]: BUTTON_THEME,
    [INPUT_KEY]: INPUT_THEME,
    colors: ColorsTheme,
    fonts: {
        base:
            'Avenir Next, Segoe UI, Helvetica Neue, Helvetica, Roboto, sans-serif',
        mono: `'SFMono-Regular', Consolas, Liberation Mono, Menlo, Courier, monospace`,
    },
    // These are all minimums
    breakpoints: {
        xs: 0,
        sm: 576,
        md: 768,
        lg: 992,
        xl: 1200,
        xxl: 1600,
    },
    // todo: we should likely remove the color.borders colors and just use this
    // or find a way to reference the colors.border colors here
    // e.g. `thin solid borders.base`
    borders: {
        light: `thin solid ${ColorsTheme.borders.light}`,
        base: `thin solid ${ColorsTheme.borders.base}`,
        dark: `thin solid ${ColorsTheme.borders.dark}`,
    },

    skeleton: {
        colorStart: '#E7E7E7',
        colorEnd: '#D3D3D3',
    },
};

// Exporting out @xstyled's ThemeProvider. Only way I could get it to work w/o issues.
// export class ThemeProvider2 extends XstyledThemeProvider {}

interface IterableObject {
    [s: number]: number | string | boolean | IterableObject;
  }

interface DebugProps {
    property: number | string | boolean | IterableObject;
    propertyName: string;
    rootProperty?: boolean;
}

interface StyledDebugProps {
    indent?: number;
}

const StyledDebug = styled('div')<StyledDebugProps>`
    font-family: mono;
    margin-top: 0.5rem;
    color: white;

    .category .category h2 {
        background-color: #ec8947;
    }

    .category .category .box {
        border-color: #ec8947;
    }

    .category .category .category h2 {
        background-color: #b84343;
        color: white;
    }

    .category .category .category .box {
        border-color: #b84343;
    }

    .category .category .category .category h2 {
        background-color: #91cdd2;
        color: black;
    }

    .category .category .category .category .box {
        border-color: #91cdd2;
    }

    .category .category .category .category .category h2 {
        background-color: #567224;
        color: white;
    }

    .category .category .category .category .category .box {
        border-color: #567224;
    }
`;

const StyledH2 = styled('h2')`
    ${spaceStyles}
    background-color: #ffea92;
    color: black;
    margin-bottom: 0;
    margin-top: 0;
    padding: 0.25rem 0.5rem 0.25rem 1rem;
    font-size: 1rem;
    cursor: pointer;
    border-radius: 0 1rem 1rem 0;
`;

interface BoxProps extends SpaceProps {
    show?: boolean;
}

const Box = styled('div')<BoxProps>`
    ${spaceStyles}
    width: 20rem;
    border-left: solid #ffea92 2px;
    margin: 0.5rem;
    ${({show}) => `display: ${show ? 'block' : 'none'}`}
`;

const Category = styled('div')<SpaceProps>`
    ${spaceStyles}
    width: 20rem;
`;


const ColorBox = styled('div')<any>`
    width: 1rem;
    height: 1rem;
    display: inline-block;
    vertical-align: middle;
    margin-left: 0.5rem;
    background-color: ${({backgroundColor}) => backgroundColor};
    border: solid white 1px;
`;

const Debug = ({property, propertyName}: DebugProps) => {
    const [show, setShow] = React.useState(false);

    return(
        <StyledDebug>
            {(property || property === 0)
                ? (
                    typeof property === 'number' ||
                    typeof property === 'string' ||
                    typeof property === 'boolean' ||
                    typeof property === 'function'
                        ? (
                                <>
                                    <span style={{fontWeight: 'bold'}}>{propertyName}: </span>
                                    <span>
                                        {typeof property === 'function'
                                            ? 'function'
                                            : /^#[a-zA-Z0-9]{4,7}/.test(property.toString())
                                                ? (<>
                                                    <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                                    <ColorBox backgroundColor={property} />
                                                    </>)
                                                : <span style={{fontStyle: 'italic'}}>{property.toString()}</span>
                                        }
                                    </span>
                                </>
                            )
                        :
                            <Category pl={0} className="category">
                                <StyledH2 onClick={() => setShow(!show)}>{propertyName}</StyledH2>
                                <Box pl={2} className="box" show={show}>
                                    {Object.values(property).map((_property, index) => (
                                        <Debug
                                            key={index}
                                            property={_property}
                                            propertyName={Object.getOwnPropertyNames(property)[index]}
                                        />
                                    ))}
                                </Box>
                            </Category>
                    )
                : (
                    <>
                       Empty.
                    </>
                )
            }
        </StyledDebug>
    );
};

const DebugWrapper = styled('div')`
    background-color: rgba(0,0,0,.75);
    border-left-color: transparent;
    position: fixed;
    left: 0;
    top: 0;
    height: 100%;
    overflow: hidden;
    width: 30rem;
    padding: 1rem 2rem 1rem 0;

    &.closed {
        height:2.4rem;
        width: auto;
        padding-right: 1rem;

        .debug-scroll {
            display: none;
        }
    }
`;

const DebugScroll = styled('div')`
    height: 100%;
    overflow-y: auto;
    width: 100%;
`;

interface ThemeProviderProps {
    children?: any;
    debug?: boolean;
    theme?: any;
}

export const ThemeProvider = ({ children, debug, theme }: ThemeProviderProps) => {
    const isOpen = localStorage.themeProviderDebug === 'true';
    const [isOpenState, setIsOpenState] = React.useState(isOpen);

    return (
        <XstyledThemeProvider theme={theme}>
            {debug &&
                <DebugWrapper className={classNames(!isOpenState ? 'closed' : null)}>
                    <Button
                        onClick={() => {
                            localStorage.themeProviderDebug = !isOpenState;
                            setIsOpenState(!isOpenState);
                        }}
                        size="sm"
                    >
                        ThemeProvider Debug
                    </Button>

                    <DebugScroll className="debug-scroll">
                        <Debug property={theme} propertyName="Theme" />
                    </DebugScroll>
                </DebugWrapper>
            }
            {children}
        </XstyledThemeProvider>
    );
};
