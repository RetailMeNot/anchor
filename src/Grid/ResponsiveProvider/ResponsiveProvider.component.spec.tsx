// REACT
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { ThemeProvider } from '@xstyled/styled-components';
// COMPONENT
import { ResponsiveProvider } from './ResponsiveProvider.component';
import { RootTheme } from '../../theme';
// ENZYME
import { shallow, mount } from 'enzyme';

jest.useFakeTimers();

describe('Component: ResponsiveProvider', () => {
    it('should be defined', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <ResponsiveProvider />
            </ThemeProvider>
        );
        const wrapper = mount(subject);
        const component = shallow(subject);

        expect(subject).toBeDefined();
        expect(wrapper).toBeDefined();
        expect(component).toBeDefined();
        const tree = renderer.create(subject).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('should use breakpoint prop if passed.', () => {
        const breakpoints = {
            lg: 300,
            sm: 100,
        };
        const subject = <ResponsiveProvider breakpoints={breakpoints} />;
        const wrapper = mount(subject);
        const responsiveProvider = wrapper.find('BasicResponsiveProvider');
        const bpVal1 = Object.values(
            responsiveProvider.state().breakpoints[0]
        )[0];
        const bpVal2 = Object.values(
            responsiveProvider.state().breakpoints[1]
        )[0];

        expect(bpVal1).toEqual(100);
        expect(bpVal2).toEqual(300);
    });

    it("should use RootTheme if no breakpoints are passed and there's no ThemeProvider.", () => {
        const subject = <ResponsiveProvider />;
        const wrapper = mount(subject);
        const responsiveProvider = wrapper.find('BasicResponsiveProvider');
        const bpVal1 = Object.values(
            responsiveProvider.state().breakpoints[0]
        )[0];
        const bpVal2 = Object.values(
            responsiveProvider.state().breakpoints[1]
        )[0];

        expect(bpVal1).toEqual(RootTheme.breakpoints.xs);
        expect(bpVal2).toEqual(RootTheme.breakpoints.sm);
    });

    it('should update state when the window is resized.', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <ResponsiveProvider />
            </ThemeProvider>
        );
        const wrapper = mount(subject);
        const responsiveProvider = wrapper.find('BasicResponsiveProvider');

        Object.defineProperty(window, 'innerWidth', {
            writable: true,
            configurable: true,
            value: 200,
        });
        window.dispatchEvent(new Event('resize'));
        jest.runAllTimers(); // Necessary because the event is debounced

        expect(responsiveProvider.state().innerWidth).toBe(200);
    });

    it('should remove window event listeners on unmount.', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <ResponsiveProvider />
            </ThemeProvider>
        );
        const wrapper = mount(subject);
        const remover = jest.spyOn(window, 'removeEventListener');

        wrapper.unmount();

        expect(remover).toHaveBeenCalled();
    });

    it('should show the debug box when debug is enabled.', () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <ResponsiveProvider debug />
            </ThemeProvider>
        );
        const wrapper = mount(subject);

        expect(wrapper.find('StyledDebug')).toBeDefined();
    });

    it("shouldn't bother doing anything with the window if it doesn't exist.", () => {
        const subject = (
            <ThemeProvider theme={RootTheme}>
                <ResponsiveProvider windowDep={false} />
            </ThemeProvider>
        );
        const wrapper = mount(subject);
        const responsiveProvider = wrapper.find('BasicResponsiveProvider');
        const remover = jest.spyOn(window, 'removeEventListener');

        wrapper.unmount();

        expect(remover).toHaveBeenCalled();
        expect(responsiveProvider.state().innerWidth).toBe(0);
    });
});
