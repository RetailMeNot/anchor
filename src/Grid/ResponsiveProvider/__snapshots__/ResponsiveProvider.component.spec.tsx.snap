// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: ResponsiveProvider should be defined 1`] = `null`;
